### PR TITLE
Enable Postgres to run on a different host than the web server

### PR DESCRIPTION
**Pull request by estadtherr**
This pull request makes the changes necessary to enable the Nominatim application to run with a Postgres server located on a separate host from the PHP application.

Most of the changes are in the `setup.php` and `update.php` scripts to ensure the setup commands are executing properly against the (possibly remote) database server. I also amped up the error handling and sub-command execution logic in those scripts as a byproduct of my work.

I updated relevant documentation as necessary, and made the changes necessary to make the `behave` unit tests run successfully against a non-local database server.

(this is a much more complete implementation of the change described in #552)

### CODE REVIEW

**Comment by lonvia** [comment]
Thank you for tackling this issue again. Can I kindly ask that you remove the coding style fixes from this PR? You are warmly welcome to resubmit them in a separate PR but mixing these kind of changes with functional code changes makes it very, very hard to review.

**Comment by estadtherr** [comment]
I reviewed the diffs again, and I believe you're referring to the deltas where my editor removed trailing white space and capitalized SQL keywords. I reworked the PR branch to remove those changes. If that's not what you were looking for, let me know!

File: utils/setup.php [hunk 1]
```diff
@@ -70,10 +72,6 @@
     $iInstances = 1;
     warn("resetting threads to $iInstances");
 }
-if ($iInstances > getProcessorCount()) {
-    $iInstances = getProcessorCount();
-    warn("resetting threads to $iInstances");
-}
```
Comment (lonvia) [inline utils/setup.php]: Unrelated.

File: utils/update.php [hunk 2]
```diff
@@ -68,11 +82,13 @@
...
     }
-    $sSetup = CONST_InstallPath.'/utils/setup.php';
-    $iRet = -1;
-    passthru($sSetup.' --create-functions --enable-diff-updates', $iRet);
-    if ($iRet != 0) {
-        fail('Error running setup script');
+    if ($aResult['update-functions']) {
+        $sSetup = CONST_InstallPath.'/utils/setup.php';
+        $iRet = -1;
+        passthru($argv[0].' '.$sSetup.' --create-functions --enable-diff-updates', $iRet);
```
Comment (lonvia) [inline utils/update.php]: Wouldn't $argv[0] be the name of the executing script? I don't see how that belongs here.

File: utils/update.php [hunk 3]
```diff
@@ -68,11 +82,13 @@
         echo "Does the URL point to a directory containing OSM update data?\n\n";
         fail('replication URL not reachable.');
     }
-    $sSetup = CONST_InstallPath.'/utils/setup.php';
-    $iRet = -1;
-    passthru($sSetup.' --create-functions --enable-diff-updates', $iRet);
-    if ($iRet != 0) {
-        fail('Error running setup script');
+    if ($aResult['update-functions']) {
```
Comment (lonvia) [inline utils/update.php]: I'd rather not have more options here.

File: data/words.sql [hunk 4]
```diff
@@ -23,6 +23,7 @@ CREATE TABLE word_frequencies (
     count bigint
 );
 
+TRUNCATE TABLE word_frequencies;
```
[the remark this answers is not in the document]
Comment (estadtherr) [reply data/words.sql]: One of the issues I encountered when iterating on making the `setup.php` script run successfully and correctly against a remote database server was determining how far to roll back the database schema and data in order to repeat one of the steps in the setup process. Many of the `sql/*.sql` files have explicit statements to support running against an existing database (`DROP TABLE IF EXISTS …` before a `CREATE TABLE`, `CREATE OR REPLACE FUNCTION …` as opposed to just `CREATE FUNCTION …`, etc.), which enables iteration on the setup. There is explicit support in the `setup.php` script for iterating on the setup as well (`CREATE EXTENSION IF NOT EXISTS …` in `--setup-db` as opposed to just `CREATE EXTENSION …`, `TRUNCATE …` and `DROP SEQUENCE …` statements in the `--load-data` section, etc.). However, the `data/words.sql` file simply loads new data without accounting for possibly existing data, which caused a conflict when repeating `setup.php`.

In essence, the change rationale was two-fold:
1. eliminate an error encountered by `setup.php`
2. make the `--load-data` step as repeatable as the rest of the steps

I'm perfectly fine removing this delta, or implementing the repeatability a different way (e.g. putting the `TRUNCATE …` in the `--load-data` section of `setup.php`), so let me know what you decide!

File: docs/admin/Faq.md [hunk 5]
```diff
@@ -112,7 +112,7 @@ to get the full error message.
 On CentOS v7 the PostgreSQL server is started with `systemd`.
 Check if `/usr/lib/systemd/system/httpd.service` contains a line `PrivateTmp=true`.
 If so then Apache cannot see the `/tmp/.s.PGSQL.5432` file. It's a good security feature,
-so use the [preferred solution](../appendix/Install-on-Centos-7/#adding-selinux-security-settings).
+so use the [[#PostgreSQL_UNIX_Socket_Location_on_CentOS|preferred solution]]
```
[the remark this answers is not in the document]
Comment (estadtherr) [reply docs/admin/Faq.md]: This change showed up when I rebased to master from v3.0.1. I agree it doesn't look right, so I'll remove it.

File: nominatim/index.c [hunk 6]
```diff
@@ -283,6 +286,23 @@ void nominatim_index(int rank_min, int rank_max, int num_threads, const char *co
...
+    num_rows = PQntuples(res);
+    if (num_rows > 0)
+    {
+        result_string = PQgetvalue(res, 0, 0);
+        if (result_string && (strlen(result_string) > 0) && (strcasecmp(result_string, "C") != 0))
+        {
+            // non-default locale if the result exists, is non-empty, and is not "C"
+            db_has_locale = 1;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (estadtherr) [reply nominatim/index.c]: Depending on the configuration of the Postgres cluster, a remote user may not be allowed to run the `SET lc_messages to 'C'` command. Since 'C' is the default locale setting (or, rather, lack thereof) on a Postgres cluster, it is often unnecessary for this setup script to change it. This change guards against that case, and doesn't try to change `LC_MESSAGES` to 'C' if that is already the configuration.

File: utils/setup.php [hunk 7]
```diff
@@ -221,9 +252,6 @@
 if ($aCMDResult['create-functions'] || $aCMDResult['all']) {
     info('Create Functions');
     $bDidSomething = true;
-    if (!file_exists(CONST_InstallPath.'/module/nominatim.so')) {
-        fail('nominatim module not built');
-    }
```
[the remark this answers is not in the document]
Comment (estadtherr) [reply utils/setup.php]: In the case where `setup.php` is running on a host other than the database server (which is likely to be the normal case when the application baseline is on a web server host and the database server is on a separate host), the `nominatim.so` file will not and should not exist on the local filesystem.

I was thinking that this check in the `--setup-db` section would be a sufficient notification that the module is absent on the database server (whether remote or local):
```    // Try accessing the C module, so we know early if something is wrong
    // and can simply error out.
    $sSQL = "CREATE FUNCTION nominatim_test_import_func(text) RETURNS text AS '";
    $sSQL .= $modulePath."/nominatim.so', 'transliteration' LANGUAGE c IMMUTABLE STRICT";
    $sSQL .= ';DROP FUNCTION nominatim_test_import_func(text);';
    $oResult = $oDB->query($sSQL);
```

File: utils/setup.php [hunk 8]
```diff
@@ -413,14 +441,23 @@
...
+                $resultError = pg_result_error($pgresult);
+                echo '-- error text ' . $i . ': ' . $resultError . '\n';
+                $failed = true;
+            }
         }
-        sleep(1);
-        echo '.';
+    }
+    if ($failed) {
+        fail('SQL errors loading placex and/or location_property_osmline tables');
```
[the remark this answers is not in the document]
Comment (estadtherr) [reply utils/setup.php]: During development of this change, my first attempt at running the `--load-data` steps failed due to errors that I later resolved. The errors were difficult to resolve because the original implementation of this block never checked the completion status of the `INSERT INTO placex …` or `insert into location_property_osmline …` statements, and never reported any errors resulting from those queries. I implemented proper status and error checking and reporting in order to discover the errors, and left that code in the PR because it was a clear improvement to the existing implementation.

File: utils/setup.php (hunk 1 above)
Comment (estadtherr) [reply utils/setup.php]: This is actually directly related. In the case where `setup.php` is running on a host other than the database server, it is incorrect for the processor count on the local host to be used to limit the number of concurrent queries allowed to run on the database server. Engineering a way for the PHP script to ask the database server for a processor count seemed excessive, not to mention a way to infer the local/remote status of the database server from the result of `DB::parseDSN(CONST_Database_DSN)`, so I chose to allow `setup.php` to trust the value provided on the command line.

File: utils/update.php (hunk 3 above)
Comment (estadtherr) [reply utils/update.php]: I wanted to account for the case where `setup.php` was run with `--enable-diff-updates`, and therefore it is unnecessary for `update.php` to run it again.

**Comment by estadtherr** [comment]
> You have also neglected to follow the naming style for variables. They should be prefixed with the variable type and camel cased.

This preference wasn't indicated in the contributor's guide, and `phpcs` did not indicate any naming issues. Can you describe your preferred style in more detail?

File: utils/update.php (hunk 2 above)
Comment (estadtherr) [reply utils/update.php]: My intent was to account for the case where the `php` interpreter was executed from a location other than `/usr/bin/php` (as stated in the interpreter directive at the top of `setup.php`). In that case, running `setup.php` by itself would fail. `$argv[0]` should reference the interpreter executable path, but now that I'm looking at it again, I think you're right that PHP populates `$argv[0]` with the script path instead of the actual system `argv[0]`.

I can revert this, unless you have a suggestion for executing the same interpreter executable that is running the `update.php` script.

File: data/words.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (lonvia) [reply data/words.sql]: `data/words.sql` not only loads data but also creates the table. So a DROP TABLE IF  EXISTS before the CREATE would be more appropriate and also guards against layout changes in the table.

File: nominatim/index.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (lonvia) [reply nominatim/index.c]: The code later detects deadlocks by parsing the error message 'ERROR:  deadlock detected'. If that message doesn't come out in English, that part of the code will fail. I wonder if it would be better to error out at this point if the locale is not in English and cannot be set. Or if you know of a less brittle way to detect the error condition, that would be even better.

File: utils/setup.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (lonvia) [reply utils/setup.php]: `--create-function` is rerun occasionally without `--setup-db`, so I think repeating the check makes sense here. How about you move the check from `--setup-db` in a helper function and then call that function again here. I wouldn't want to remove the check from `--setup-db` either because you wouldn't want to do the full osm2pgsql import before realising that something is wrong with the nominatim.so.

File: utils/setup.php (hunk 1 above)
Comment (lonvia) [reply utils/setup.php]: Makes sense.

File: utils/update.php (hunk 2 above)
Comment (lonvia) [reply utils/update.php]: PHP_BINARY would be the way to go but this is really something for a different PR. I suspect that there are other places that need to be adapted.

File: utils/update.php (hunk 3 above)
Comment (lonvia) [reply utils/update.php]: I'd be okay with an option `--no-update-functions` if there is really an important use case but the default should be that `setup.php` is executed. I wouldn't consider speed an important use case, though. Rather something along the lines of "I have to put my external server into a special mode to be able to run the --create-functions stuff and don't want to have to do that again".

**Comment by lonvia** [comment]
> This preference wasn't indicated in the contributor's guide, and phpcs did not indicate any naming issues. Can you describe your preferred style in more detail?

Yeah, I noticed too that this was missing in the contributor's guide. Sorry about that I've added a note about variable naming now. Note that this only applies to PHP code. Python uses standard underscore notation. (Historical reasons, etc.) The rule of thumb is to stick as close as possible to whatever style the surrounding code has.

github now reports conflicts, do you mind rebasing to latest master?

**Comment by estadtherr** [comment]
I rebased to the latest master, and made the variable naming changes. I'm continuing to look at your other comments, so more changes will be coming.

File: nominatim/index.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (estadtherr) [reply nominatim/index.c]: I hadn't noticed that deadlock detection logic. I don't know of a more reliable way to ascertain whether a failed query was the result of the server's built in deadlock detection algorithm. However, my change still supports the existing logic: in essence, the new logic is "Set LC_MESSAGES to 'C' **only if it isn't already**".

File: utils/update.php (hunk 2 above)
Comment (estadtherr) [reply utils/update.php]: Agreed that a separate PR makes the most sense. I'll take it out of this one.

File: utils/update.php (hunk 3 above)
Comment (estadtherr) [reply utils/update.php]: In our case, our web service platform runs the `php` interpreter from a non-standard location (see other comment around this change), so running `…/setup.php` as the command would fail. We therefore always run the initial `setup.php` with the `--enable-diff-updates` argument (with the understanding that the `placex_insert` triggers will be slower). `--no-update-functions` will solve that, though, so I'll go ahead and make that change.

File: utils/setup.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (estadtherr) [reply utils/setup.php]: I refactored it as you described.

**Comment by lonvia** [comment]
I just noticed that the module path is handed in via a command line option. I would have a slight preference for making it a configuration variable settable via local.php. The reason is that there are occasions where `utils/setup.php --create-functions` needs to be rerun, for example, when upgrading to a newer Nominatim version and then the additional option is easily forgotten.

I've opened a separate issue #1109 for supporting custom PHP binaries.

File: data/words.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (estadtherr) [reply data/words.sql]: The `DROP TABLE IF EXISTS …` option might be best. I'll go ahead and make that change to `data/words.sql`. That will at least make `--load-data` repeatable. The other SQL files in `data/` that are loaded in the `--setup-db` section follow the same pattern, but there are other things in the `--setup-db` section that would need to be addressed to make it repeatable.

**Comment by estadtherr** [comment]
> I just noticed that the module path is handed in via a command line option. I would have a slight preference for making it a configuration variable settable via local.php. The reason is that there are occasions where utils/setup.php --create-functions needs to be rerun, for example, when upgrading to a newer Nominatim version and then the additional option is easily forgotten.

I considered that, but ended up deliberately going the command line route with the goal of keeping one-time database setup configuration out of the application's runtime configuration file. If you still would rather go the `local.php` route, I can certainly make that change, but wanted to at least share my reasoning beforehand. Let me know!

**Comment by lonvia** [comment]
> I considered that, but ended up deliberately going the command line route with the goal of keeping one-time database setup configuration out of the application's runtime configuration file.

It's a bit of an in-between thing. It's true that the path itself is only used in setup.php but the file needs to stay in that location during runtime. Experience tells me that it is better then to have it somewhere written down where you've put the file. So, yes, I'd prefer the `local.php` route.

**Comment by estadtherr** [comment]
> It's a bit of an in-between thing. It's true that the path itself is only used in setup.php but the file needs to stay in that location during runtime. Experience tells me that it is better then to have it somewhere written down where you've put the file. So, yes, I'd prefer the local.php route.

I made that change. I think that’s the last outstanding comment. Let me know if you have more questions or would like any additional changes!

**Comment by lonvia** [comment]
Merged as is. Thank you for your contribution and a special thanks for patiently putting up with all my nitpicking. ;)

**Comment by estadtherr** [comment]
Great, thanks! I was going to suggest merging as is, since then the commits line up with the review history. And no problem at all with the "nitpicking" - I'd be doing the same thing in your shoes! :)

**Comment by olablt** [comment]
hey, i think my problem is related to this update

I have my web server and postgres database on different docker containers

when I run this command from web server (Centos7 container)

```
sudo -u nominatim psql -h postgis -lqt
```

I get the list of the tables on the host `postgis`

but when I run 

```
sudo -u nominatim ./utils/setup.php --osm-file /app/data/lithuania-latest.osm.pbf --all -v
```

I get the error `could not connect to server: No such file or directory`

my `local.php`

```
<?php
 // General settings
 @define('CONST_Database_DSN', 'pgsql://nominatim@' . getenv('PGHOST') . ':5432/nominatim'); // <driver>://<username>:<password>@<host>:<port>/<database>
 @define('CONST_Website_BaseURL', '/');
 // Software versions
 @define('CONST_Postgresql_Version', '9.6');
 @define('CONST_Postgis_Version', '2.3');
```

did I missed something?

**Comment by estadtherr** [comment]
Can you show more of the output from `setup.php`? The only thing that looks suspicious is that you're not providing a database password in the DSN. Are you supplying the password with a `$HOME/.pgpass` file?

**Comment by lonvia** [comment]
Please open a new issue instead of commenting on closed PRs.

**Comment by phlegx** [comment]
May I ask what the status with this issue is? Basically the main issue still for to be able to run Postgres in a different container is still this check in setup.php:

```
if (!file_exists(CONST_InstallPath.'/module/nominatim.so')) {
  fail('nominatim module not built');
}
```
commenting this check out allows me to run 

```./src/build/utils/setup.php --osm-file /app/src/downloads/dach-latest.osm.pbf --all --threads "$IMPORT_THREADS"```

With Nominatim and Postgres in two different containers.

**Comment by mtmail** [comment]
@phlegx Please open a new issue, this PR got merged last year. I may seem extra work but comments on closed issues get lost easily. The web server needs `nominatim.so` for string normalization during import, the database server needs `nominatim.so` for string normalization during querying.

**Comment by phlegx** [comment]
@mtmail ok thanks for the information! I will open a new issue.
